src/Grpc.Net.Client/Internal/GrpcCall.cs

```diff
@@ -835,6 +835,22 @@ private bool FinishCall(HttpRequestMessage request, bool diagnosticSourceEnabled
             return true;
         }
 
+#if NET5_0_OR_GREATER
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:UnrecognizedReflectionPattern",
+            Justification = "The values being passed into Write have the commonly used properties being preserved with DynamicDependency.")]
```

Comment:
The `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)]` annotation will keep properties on the `TValue`, but it will not do that recursively. Diagnostic source might recourse into properties of the properties. See some discussion in https://github.com/dotnet/runtime/issues/50454.

I don't think this code would still work as intended if the trimmer were to trim all properties on `HttpRequestMessage` (which is one of the property types in question in this file). There's nothing preventing that here.

Reply:
That's true, I don't think there is anything else that can be done.

Reply:
(btw I copied this technique from dotnet/runtime so I'm guessing the code was inspired by that issue or vice versa)

Reply:
I think in the dotnet/runtime repo we ensure the properties on nested things are preserved through other means, if there are any. We were pretty careful not to suppress warnings that could result in different behavior after trimming.

The warnings are the only thing between "the app works as intended after trimming" and "the app is broken after I finished all my debugging, I've decided to publish a release build, and none of my unit tests caught it".

The annoying thing about trimming is that it cannot be unit tested (unit testing changes the set of things that are kept) so all trimming related failures only show up once one does E2E testing. E2E testing for many apps means "manually going through many scenarios". Trimming warnings are the only line of defense we have and an invalid suppression often ends up being a disservice to the customer.

Reply:
Check out what we did in `System.Net.Http.DiagnosticsHandler` in https://github.com/dotnet/runtime/pull/50265.

We added `DynamicDependency` attributes for the important properties (i.e. the ones that are automatically selected by [Microsoft.Diagnostics.Monitoring.EventPipe](https://github.com/dotnet/diagnostics/blob/ffd0254da3bcc47847b1183fa5453c0877020abd/src/Microsoft.Diagnostics.Monitoring.EventPipe/Configuration/HttpRequestSourceConfiguration.cs#L36-L40)). For example:

https://github.com/dotnet/runtime/blob/395038186623985b48a1ac1ccb06a77291f2f2ed/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L218-L226

https://github.com/dotnet/runtime/blob/395038186623985b48a1ac1ccb06a77291f2f2ed/src/libraries/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs#L296-L299

This ensures the trimmer won't remove these properties from `HttpRequestMessage`, `HttpResponseMessage`, or even `Uri`.

Unfortunately DiagnosticSource's design is intrinsically incompatible with trimming. We make a "best effort" to ensure the important properties are kept. And we added to the error message that gets logged when a property cannot be found at runtime:

> "If you published the application with PublishTrimmed=true, ensure the property was not trimmed away."

So if a user is trying to get a property that was trimmed, the developer can explicitly add it to the list of properties to be kept.

Reply:
Better? [`0218ab6` (#1465)](https://github.com/grpc/grpc-dotnet/pull/1465/commits/0218ab6c04b4de45f4abeb994474c5b8b8f83da5)

I understand it's not perfect.